crypto/threads_common.c

```diff
@@ -0,0 +1,451 @@
+ *          │        .              cast to uintptr_t
+ *          │        .
+ *          │        .
+ *          │┌───────────────┐
+ *          ││        ┌──────┼────► ┌─────────────────┐
+ *          ││        │      │      │                 │
+ *          └┴────────┴──────┘      └─────────────────┘
+ *                                  per-<thread*ctx> data
+ * It uses sparse arrays to map:
+ *   - Thread-local key IDs to master key entries.
+ *   - Library context pointers to context-specific data.
+ *
+ * Two primary functions are provided:
+ *   - CRYPTO_THREAD_get_local_ex() retrieves data associated with a key and
+ *     context.
+ *   - CRYPTO_THREAD_set_local_ex() associates data with a given key and
+ *     context, allocating tables as needed.
+ *
+ * Internal structures:
+ *   - CTX_TABLE_ENTRY: wraps a context-specific data pointer.
+ *   - MASTER_KEY_ENTRY: maintains a table of CTX_TABLE_ENTRY and an optional
+ *     cleanup function.
+ *
+ * The implementation ensures:
+ *   - Lazy initialization of master key data using CRYPTO_ONCE.
+ *   - Automatic cleanup of all context and key mappings when a thread exits.
+ *
+ * Cleanup routines:
+ *   - clean_ctx_entry: releases context-specific entries.
+ *   - clean_master_key_id: releases all entries for a specific key ID.
+ *   - clean_master_key: top-level cleanup for the thread-local master key.
+ *
+ */
+
+#include <openssl/crypto.h>
+#include <crypto/cryptlib.h>
+#include <crypto/sparse_array.h>
+#include "internal/cryptlib.h"
+#include "internal/threads_common.h"
+
+/**
+ * @struct CTX_TABLE_ENTRY
+ * @brief Represents a wrapper for context-specific data.
+ *
+ * This structure is used to hold a pointer to data that is associated
+ * with a particular `OSSL_LIB_CTX` instance in a thread-local context
+ * mapping. It is stored within a sparse array, allowing efficient
+ * per-context data lookup keyed by a context identifier.
+ *
+ * @var CTX_TABLE_ENTRY::ctx_data
+ * Pointer to the data associated with a given library context.
+ */
+typedef struct ctx_table_entry {
+    void *ctx_data;
+} CTX_TABLE_ENTRY;
+
+/*
+ * define our sparse array of CTX_TABLE_ENTRY functions
+ */
+DEFINE_SPARSE_ARRAY_OF(CTX_TABLE_ENTRY);
+
+/**
+ * @struct MASTER_KEY_ENTRY
+ * @brief Represents a mapping of context-specific data for a TLS key ID.
+ *
+ * This structure manages a collection of `CTX_TABLE_ENTRY` items, each
+ * associated with a different `OSSL_LIB_CTX` instance. It supports
+ * cleanup of stored data when the thread or key is being destroyed.
+ *
+ * @var MASTER_KEY_ENTRY::ctx_table
+ * Sparse array mapping `OSSL_LIB_CTX` pointers (cast to uintptr_t) to
+ * `CTX_TABLE_ENTRY` structures that hold context-specific data.
+ *
+ */
+typedef struct master_key_entry {
+    SPARSE_ARRAY_OF(CTX_TABLE_ENTRY) *ctx_table;
+} MASTER_KEY_ENTRY;
+
+/*
+ * Define our sparse array functions for MASTER_KEY_ENTRY structs
+ */
+DEFINE_SPARSE_ARRAY_OF(MASTER_KEY_ENTRY);
```

Comment:
I think this should just be a regular array

Reply:
The FIPS provider manages its own thread local data completely separately to libcrypto. The size of the array can be different between the two because they are entirely independent. So if we have a newer libcrypto which no longer users a specific key id that the fips provider uses - it makes no difference. The fips provider is entirely separate anyway.

Reply:
Sparse arrays are not going to save space when only a few indices are missing.  They are implemented as a tree of blocks and any gap under the size of a block will use a full block.

Access is O(log n) where n is the number of elements in the sparse array.  The block size is fairly large so this isn't consequential most of the time.